File: extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/runtime/KotlinJpaOperations.java
package io.quarkus.hibernate.orm.panache.kotlin.runtime;

import java.util.List;
import java.util.stream.Stream;

import javax.persistence.EntityManager;

import io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations;

public class KotlinJpaOperations extends AbstractJpaOperations<PanacheQueryImpl<?>> {
    /**
     * Provides the default implementations for quarkus to wire up. Should not be used by third party developers.
     */
    public static final KotlinJpaOperations INSTANCE = new KotlinJpaOperations();

    @Override
    protected PanacheQueryImpl<?> createPanacheQuery(EntityManager em, String query, String orderBy,
            Object paramsArrayOrMap) {
        return new PanacheQueryImpl<>(em, query, orderBy, paramsArrayOrMap);
    }

    @Override
    public List<?> list(PanacheQueryImpl<?> query) {
        return query.list();
    }

    @Override
    public Stream<?> stream(PanacheQueryImpl<?> query) {
        return query.stream();
    }
}

File: extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/orm/panache/kotlin/runtime/KotlinJpaOperations.kt
package io.quarkus.hibernate.orm.panache.kotlin.runtime

import io.quarkus.hibernate.orm.panache.common.runtime.AbstractManagedJpaOperations
import io.quarkus.panache.common.Sort
import org.hibernate.Session

class KotlinJpaOperations : AbstractManagedJpaOperations<PanacheQueryImpl<*>>() {
    override fun createPanacheQuery(
        session: Session,
        entityClass: Class<*>,
        hqlQuery: String,
        originalQuery: String?,
        sort: Sort?,
        paramsArrayOrMap: Any?,
    ) = PanacheQueryImpl<Any>(session, entityClass, hqlQuery, originalQuery, sort, paramsArrayOrMap)

    override fun list(query: PanacheQueryImpl<*>) = query.list()

    override fun stream(query: PanacheQueryImpl<*>) = query.stream()

    companion object {
        /**
         * Provides the default implementations for quarkus to wire up. Should not be used by third
         * party developers.
         */
        @JvmField val INSTANCE = KotlinJpaOperations()
    }
}
